diff --git a/accel/tcg/tcg-runtime-sym.c b/accel/tcg/tcg-runtime-sym.c
index df709fb..af73d67 100644
--- a/accel/tcg/tcg-runtime-sym.c
+++ b/accel/tcg/tcg-runtime-sym.c
@@ -600,24 +600,24 @@ static void *sym_movcond_internal(CPUArchState *env,
                               int32_t comparison_operator,
                               uint64_t is_taken, uint8_t result_bits)
 {
-    if (c1_expr == NULL && c2_expr == NULL && v1_expr == NULL && v2_expr == NULL) {
-        return NULL;
+    if (c1_expr == NULL && c2_expr == NULL) {
+        return is_taken ? v1_expr : v2_expr;
     }
 
     if (c1_expr == NULL) {
-        c1_expr = _sym_build_integer(c1, _sym_bits_helper(c2_expr));
+        c1_expr = _sym_build_integer(c1, result_bits);
     }
 
     if (c2_expr == NULL) {
-        c2_expr = _sym_build_integer(c2, _sym_bits_helper(c1_expr));
+        c2_expr = _sym_build_integer(c2, result_bits);
     }
 
     if (v1_expr == NULL) {
-        v1_expr = _sym_build_integer(v1, _sym_bits_helper(c1_expr));
+        v1_expr = _sym_build_integer(v1, result_bits);
     }
 
     if (v2_expr == NULL) {
-        v2_expr = _sym_build_integer(v2, _sym_bits_helper(c1_expr));
+        v2_expr = _sym_build_integer(v2, result_bits);
     }
 
     assert(_sym_bits_helper(c1_expr) == result_bits);
@@ -633,7 +633,8 @@ static void *sym_movcond_internal(CPUArchState *env,
     assert(_sym_bits_helper(condition_ext) == result_bits);
 
     void *v1_masked = _sym_build_and(v1_expr, condition_ext);
-    void *v2_masked = _sym_build_and(v2_expr, condition_ext);
+    void *v2_masked = _sym_build_and(v2_expr,
+        _sym_build_xor(condition_ext, _sym_build_integer(UINT64_MAX, result_bits)));
 
     return _sym_build_xor(v1_masked, v2_masked);
 }
